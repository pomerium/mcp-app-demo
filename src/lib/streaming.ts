export function streamText(
  answer: AsyncIterable<any>,
  onMessageId?: (messageId: string) => void,
): Response {
  const encoder = new TextEncoder()
  const messageId = `msg-${Math.random().toString(36).slice(2)}`

  const stream = new ReadableStream({
    async start(controller) {
      // Send initial message ID
      controller.enqueue(
        encoder.encode(
          `f:${JSON.stringify({
            messageId,
          })}\n`,
        ),
      )

      if (onMessageId) {
        onMessageId(messageId)
      }

      let buffer = ''
      let reasoningSummaryBuffer = ''

      const flush = () => {
        if (buffer) {
          controller.enqueue(encoder.encode(`0:${JSON.stringify(buffer)}\n`))
          buffer = ''
        }
      }

      for await (const chunk of answer) {
        // Handle tool-related chunks
        switch (chunk.type) {
          case 'response.mcp_list_tools.in_progress':
            controller.enqueue(
              encoder.encode(
                `t:${JSON.stringify({
                  type: 'tool_in_progress',
                  itemId: chunk.item_id,
                })}\n`,
              ),
            )
            break
          case 'response.mcp_list_tools.completed':
            controller.enqueue(
              encoder.encode(
                `t:${JSON.stringify({
                  type: 'tool_completed',
                  itemId: chunk.item_id,
                })}\n`,
              ),
            )
            break
          case 'response.output_item.done':
            if (chunk.item.type === 'mcp_list_tools') {
              controller.enqueue(
                encoder.encode(
                  `t:${JSON.stringify({
                    type: 'tool_done',
                    serverLabel: chunk.item.server_label,
                    tools: chunk.item.tools,
                  })}\n`,
                ),
              )
            }
            break
          case 'response.output_text.delta':
            if (typeof chunk.delta === 'string') {
              buffer += chunk.delta

              // Flush on sentence boundaries or when buffer gets large
              if (buffer.length > 40 || /[.!?\n]$/.test(chunk.delta)) {
                flush()
              }
            }
            break

          case 'response.content_part.added':
          case 'response.content_part.done':
            if (chunk.part?.type === 'output_text' && chunk.part.text) {
              buffer += chunk.part.text
              flush()
            }
            break

          case 'response.reasoning.delta':
            if (typeof chunk.delta === 'string') {
              controller.enqueue(
                encoder.encode(
                  `t:${JSON.stringify({
                    type: 'reasoning',
                    effort: chunk.effort,
                    summary: chunk.delta,
                    model: chunk.model,
                    serviceTier: chunk.service_tier,
                    temperature: chunk.temperature,
                    topP: chunk.top_p,
                  })}\n`,
                ),
              )
            }
            break

          case 'response.created':
          case 'response.in_progress':
            if (chunk.response?.reasoning) {
              controller.enqueue(
                encoder.encode(
                  `t:${JSON.stringify({
                    type: 'reasoning',
                    effort: chunk.response.reasoning.effort,
                    summary: chunk.response.reasoning.summary,
                  })}\n`,
                ),
              )
            }
            break

          case 'response.mcp_call.failed':
            console.error('[TOOL CALL FAILED]', chunk)

            controller.enqueue(
              encoder.encode(
                `t:${JSON.stringify({
                  type: 'mcp_call_failed',
                  itemId: chunk.item_id,
                })}\n`,
              ),
            )
            break

          case 'response.mcp_call.created':
            controller.enqueue(
              encoder.encode(
                `t:${JSON.stringify({
                  type: 'mcp_call_created',
                  itemId: chunk.item_id,
                  toolName: chunk.tool.name,
                  arguments: chunk.tool.input,
                })}\n`,
              ),
            )
            break

          case 'response.mcp_call.in_progress':
            controller.enqueue(
              encoder.encode(
                `t:${JSON.stringify({
                  type: 'mcp_call_in_progress',
                  itemId: chunk.item_id,
                })}\n`,
              ),
            )
            break

          case 'response.mcp_call_arguments.delta':
            controller.enqueue(
              encoder.encode(
                `t:${JSON.stringify({
                  type: 'mcp_call_arguments_delta',
                  itemId: chunk.item_id,
                  delta: chunk.delta,
                })}\n`,
              ),
            )
            break

          case 'response.mcp_call_arguments.done':
            controller.enqueue(
              encoder.encode(
                `t:${JSON.stringify({
                  type: 'mcp_call_arguments_done',
                  itemId: chunk.item_id,
                  arguments: chunk.arguments,
                })}\n`,
              ),
            )
            break

          case 'response.mcp_call.completed':
            controller.enqueue(
              encoder.encode(
                `t:${JSON.stringify({
                  type: 'mcp_call_completed',
                  itemId: chunk.item_id,
                })}\n`,
              ),
            )
            break

          case 'response.output_item.added':
            if (chunk.item.type === 'mcp_call') {
              controller.enqueue(
                encoder.encode(
                  `t:${JSON.stringify({
                    type: 'mcp_call',
                    itemId: chunk.item.id,
                    toolName: chunk.item.name,
                    serverLabel: chunk.item.server_label,
                    arguments: chunk.item.arguments,
                  })}\n`,
                ),
              )
            }
            break

          case 'response.reasoning_summary_text.delta':
            if (typeof chunk.delta === 'string') {
              reasoningSummaryBuffer += chunk.delta
            }
            break

          case 'response.reasoning_summary_text.done':
            if (reasoningSummaryBuffer) {
              controller.enqueue(
                encoder.encode(
                  `t:${JSON.stringify({
                    type: 'reasoning',
                    effort: chunk.effort,
                    summary: reasoningSummaryBuffer,
                    model: chunk.model,
                    serviceTier: chunk.service_tier,
                    temperature: chunk.temperature,
                    topP: chunk.top_p,
                  })}\n`,
                ),
              )
              reasoningSummaryBuffer = ''
            }
            break

          default:
            break
        }
      }

      // Flush any remaining content
      flush()
      controller.close()
    },
  })

  return new Response(stream, {
    headers: {
      'Content-Type': 'text/event-stream',
      'Cache-Control': 'no-cache',
      Connection: 'keep-alive',
    },
  })
}
