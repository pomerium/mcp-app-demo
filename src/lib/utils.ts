import { type ClassValue, clsx } from 'clsx'
import { twMerge } from 'tailwind-merge'

export function cn(...inputs: ClassValue[]) {
  return twMerge(clsx(inputs))
}

export function formatTimestamp(date: Date): string {
  return new Intl.DateTimeFormat('en-US', {
    hour: 'numeric',
    minute: 'numeric',
    hour12: true,
  }).format(date)
}

export function delay(ms: number): Promise<void> {
  return new Promise((resolve) => setTimeout(resolve, ms))
}
