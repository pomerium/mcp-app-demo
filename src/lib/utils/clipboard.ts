// Utility to copy text to clipboard, with iOS fallback
// Returns a boolean indicating success
export async function copyToClipboard(text: string): Promise<boolean> {
  let copied = false
  if (navigator.clipboard && window.isSecureContext) {
    try {
      await navigator.clipboard.writeText(text)
      copied = true
    } catch {}
  }
  if (!copied) {
    const el = document.createElement('textarea')
    el.value = text
    el.setAttribute('readonly', '')
    el.style.position = 'absolute'
    el.style.left = '-9999px'
    document.body.appendChild(el)
    el.select()
    el.setSelectionRange(0, el.value.length)
    try {
      document.execCommand('copy')
      copied = true
    } catch {}
    document.body.removeChild(el)
  }
  return copied
}
