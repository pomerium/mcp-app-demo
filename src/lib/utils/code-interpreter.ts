// Utility functions for file handling (shared between components)

export interface AnnotatedFile {
  type: string
  container_id: string
  file_id: string
  filename: string
  start_index?: number
  end_index?: number
}

const reImageFileExtension = /\.(jpg|jpeg|png|gif|bmp|webp|svg)$/i

export function isImageFile(filename: string): boolean {
  return reImageFileExtension.test(filename)
}

export function createAnnotatedFileUrl(file: AnnotatedFile): string {
  const url = new URL('/api/container-file', location.origin)
  url.searchParams.set('containerId', file.container_id)
  url.searchParams.set('fileId', file.file_id)
  return url.toString()
}
