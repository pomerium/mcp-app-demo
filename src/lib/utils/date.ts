/**
 * Get the current timestamp in ISO string format
 * @returns The current timestamp in ISO string format
 */
export const getTimestamp = () => {
  return new Date().toISOString()
}
