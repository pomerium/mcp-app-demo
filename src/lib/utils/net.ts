import { createHash } from 'crypto'

/**
 * Create an ETag for a given key
 *
 * This is a simple MD5-based ETag generator for key-based caching.
 * Suitable for ephemeral content like OpenAI container files where
 * we generate ETags from metadata (containerId-fileId) rather than
 * actual file content.
 *
 * For RFC 7232 compliant ETags or content-based hashing, consider
 * using the `etag` npm package instead.
 *
 * @param key - The key to create an ETag for
 * @returns The ETag (without quotes - add quotes when setting HTTP headers)
 */
export function createEtag(key: string) {
  // MD5 is fine for ephemeral containers and key-based ETags
  return createHash('md5').update(key).digest('hex')
}
