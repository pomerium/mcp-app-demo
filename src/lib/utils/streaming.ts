/**
 * Stops a response's readable stream from being processed further.
 * This is useful for halting streaming responses when needed.
 *
 * @param response The response object to modify.
 */
export function stopStreamProcessing(response: Response) {
  const emptyStream = new ReadableStream({
    start(controller) {
      controller.close()
    },
  })

  Object.defineProperty(response, 'body', {
    value: emptyStream,
  })
}
