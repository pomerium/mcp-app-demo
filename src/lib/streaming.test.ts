import { describe, expect, it } from 'vitest'
import { stopStreamProcessing } from './utils/streaming'
import { streamText } from './streaming'

function iterableFromArray<T>(arr: Array<T>): AsyncIterable<T> {
  return {
    [Symbol.asyncIterator]() {
      let i = 0
      return {
        // eslint-disable-next-line @typescript-eslint/require-await
        next: async () =>
          i < arr.length
            ? { value: arr[i++], done: false }
            : { value: undefined, done: true },
      }
    },
  }
}

describe('stopStreamProcessing', () => {
  it('replaces response body with an empty closed stream', async () => {
    const originalStream = new ReadableStream({
      start(controller) {
        controller.enqueue('some data')
        controller.close()
      },
    })

    const response = new Response(originalStream, {
      headers: { 'Content-Type': 'text/plain' },
    })

    expect(response.body).toBe(originalStream)
    stopStreamProcessing(response)

    expect(response.body).not.toBe(originalStream)
    expect(response.body).toBeInstanceOf(ReadableStream)

    const reader = response.body?.getReader()
    expect(reader).toBeDefined()

    if (reader) {
      const result = await reader.read()
      expect(result.done).toBe(true)
      expect(result.value).toBeUndefined()
    }
  })

  it('preserves other response properties', () => {
    const originalHeaders = new Headers({ 'Content-Type': 'application/json' })
    const response = new Response('content', {
      status: 200,
      statusText: 'OK',
      headers: originalHeaders,
    })

    stopStreamProcessing(response)

    expect(response.status).toBe(200)
    expect(response.statusText).toBe('OK')
    expect(response.headers.get('Content-Type')).toBe('application/json')
  })

  it('works with responses that have no initial body', () => {
    const response = new Response(null)

    expect(() => stopStreamProcessing(response)).not.toThrow()

    expect(response.body).toBeInstanceOf(ReadableStream)
  })
})

describe('streamText', () => {
  it('emits stream_done when the stream ends', async () => {
    const chunks = [
      { type: 'response.output_text.delta', delta: "I'm glad you asked!" },
      {
        type: 'response.output_text.delta',
        delta: ' Here are a few universally nice things that',
      },
      {
        type: 'response.output_text.delta',
        delta: ' could have happened today:\n\n',
      },
      {
        type: 'response.output_text.delta',
        delta: '- Someone smiled at a stranger, brightening',
      },
      { type: 'response.output_text.delta', delta: ' their day\n' },
      {
        type: 'response.output_text.delta',
        delta: '- A teacher helped a student understand a',
      },
      { type: 'response.output_text.delta', delta: ' difficult concept\n' },
      {
        type: 'response.output_text.delta',
        delta: '- A kind person paid for someoneâ€™s coffee',
      },
      { type: 'response.output_text.delta', delta: ' in line\n' },
      {
        type: 'response.output_text.delta',
        delta: '- A pet reunited with its owner at a local',
      },
      { type: 'response.output_text.delta', delta: ' animal shelter\n' },
      {
        type: 'response.output_text.delta',
        delta: '- A friend reached out just to say hello\n\n',
      },
      {
        type: 'response.output_text.delta',
        delta: 'Would you like to hear some real, uplifting',
      },
      { type: 'response.output_text.delta', delta: ' news from today?' },
      { type: 'response.output_text.delta', delta: ' Just let me know!' },
      {
        type: 'response.tool_call_completed',
        response: {
          /* ...omitted... */
        },
      },
      {
        type: 'response.completed',
        response: {
          /* ...omitted... */
        },
      },
    ]
    const response = streamText(iterableFromArray(chunks))
    const reader = response.body!.getReader()
    let result = ''
    let done = false
    while (!done) {
      const { value, done: d } = await reader.read()
      if (value) result += new TextDecoder().decode(value)
      done = d
    }
    expect(result).toMatch(/t:{"type":"stream_done"}/)
  })

  it('emits stream_done when an error occurs', async () => {
    const consoleErrorSpy = vi
      .spyOn(console, 'error')
      .mockImplementation(() => {})

    const errorIterable = {
      [Symbol.asyncIterator]() {
        return {
          next: () => {
            throw new Error('Simulated streaming error')
          },
        }
      },
    }
    const response = streamText(errorIterable)
    const reader = response.body!.getReader()
    let result = ''
    let done = false
    while (!done) {
      const { value, done: d } = await reader.read()
      if (value) result += new TextDecoder().decode(value)
      done = d
    }

    expect(result).toMatch(/t:{"type":"stream_done"}/)
    expect(result).toMatch(/e:{"type":"error".*}/)
    expect(consoleErrorSpy).toHaveBeenCalledWith(
      'Error during streamed response:',
      expect.any(Error),
    )
    expect(consoleErrorSpy.mock.calls[0][1].message).toBe(
      'Simulated streaming error',
    )

    consoleErrorSpy.mockRestore()
  })
})
