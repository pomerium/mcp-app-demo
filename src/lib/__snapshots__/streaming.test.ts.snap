// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`streamText > emits stream_done when an error occurs 1`] = `
"f:{"messageId":"msg-2"}
e:{"type":"error","message":"An unexpected error occurred."}
t:{"type":"stream_done"}
"
`;

exports[`streamText > emits stream_done when the stream ends 1`] = `
"f:{"messageId":"msg-1"}
0:"I'm glad you asked!"
0:" Here are a few universally nice things that"
0:" could have happened today:\\n\\n"
0:"- Someone smiled at a stranger, brightening"
0:" their day\\n"
0:"- A teacher helped a student understand a"
0:" difficult concept\\n"
0:"- A kind person paid for someoneâ€™s coffee"
0:" in line\\n"
0:"- A pet reunited with its owner at a local"
0:" animal shelter\\n"
0:"- A friend reached out just to say hello\\n\\n"
0:"Would you like to hear some real, uplifting"
0:" news from today?"
0:" Just let me know!"
t:{"type":"tool_call_completed","response":{}}
t:{"type":"stream_done"}
"
`;

exports[`streamText > passes through response.created events 1`] = `
"f:{"messageId":"msg-3"}
t:{"type":"response.created","response":{"id":"resp_123","model":"gpt-4","created":1234567890}}
t:{"type":"stream_done"}
"
`;
